src/device-manager/java/src/nl/Weave/DataManagement/GenericTraitUpdatableDataSinkImpl.java

```diff
@@ -0,0 +1,444 @@
+        if (mTraitInstancePtr == 0)
+        {
+            Log.e(TAG, "unexpected err, mTraitInstancePtr is 0");
+            return null;
+        }
+        return getBytes(mTraitInstancePtr, path);
+    }
+
+    @Override
+    public long getVersion()
+    {
+        if (mTraitInstancePtr == 0)
+        {
+            Log.e(TAG, "unexpected err, mTraitInstancePtr is 0");
+            return 0;
+        }
+        return getVersion(mTraitInstancePtr);
+    }
+
+    @Override
+    public void beginRefreshData()
+    {
+        if (mTraitInstancePtr == 0)
+        {
+            Log.e(TAG, "unexpected err, mTraitInstancePtr is 0");
+            return;
+        }
+        beginRefreshData(mTraitInstancePtr);
+    }
+
+    @Override
+    public CompletionHandler getCompletionHandler()
+    {
+        return mCompHandler;
+    }
+
+    @Override
+    public void setCompletionHandler(CompletionHandler compHandler)
+    {
+        mCompHandler = compHandler;
+    }
+
+    private void onError(Throwable err)
+    {
+        if (mCompHandler == null)
+        {
+            Log.e(TAG, "unexpected err, mCompHandler is null");
+            return;
+        }
+        mCompHandler.onError(err);
+    }
+
+    private void onRefreshDataComplete()
+    {
+        if (mCompHandler == null)
+        {
+            Log.e(TAG, "unexpected err, mCompHandler is null");
+            return;
+        }
+        mCompHandler.onRefreshDataComplete();
+    }
+
+    // ----- Protected Members -----
+
+    protected CompletionHandler mCompHandler;
+
+    protected void finalize() throws Throwable
+    {
+        super.finalize();
+        close();
+    }
+
+    // ----- Private Members -----
+
+    private long mTraitInstancePtr;
+    private WDMClientImpl mWDMClient;
+    private final static String TAG = GenericTraitUpdatableDataSink.class.getSimpleName();
+
+    static {
+        System.loadLibrary("WeaveDeviceManager");
+    }
+
+    private native void init(long genericTraitUpdatableDataSinkPtr);
+    private native void close(long genericTraitUpdatableDataSinkPtr);
+    private native void beginRefreshData(long genericTraitUpdatableDataSinkPtr);
+    private native void setInt(long genericTraitUpdatableDataSinkPtr, String path, long value, boolean isConditional, boolean isSigned);
+    private native void setDouble(long genericTraitUpdatableDataSinkPtr, String path, double value, boolean isConditional);
+    private native void setBoolean(long genericTraitUpdatableDataSinkPtr, String path, boolean value, boolean isConditional);
+    private native void setString(long genericTraitUpdatableDataSinkPtr, String path, String value, boolean isConditional);
+    private native void setNULL(long genericTraitUpdatableDataSinkPtr, String path, boolean isConditional);
+    private native void setBytes(long genericTraitUpdatableDataSinkPtr, String path, byte[] value, boolean isConditional);
+    private native long getInt(long genericTraitUpdatableDataSinkPtr, String path);
```

Comment:
Can we rename the native method to `getLong` since it returns a `long`?

Reply:
agree. updated the naming on it. thanks